Python/optimizer_analysis.c

```diff
@@ -533,6 +533,128 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
+ * This calculation is essentially a traversal over the tree of function
+ * calls in this trace.
+ */
+static void
+combine_stack_space_checks(
+    _PyUOpInstruction *buffer,
+    int buffer_size
+)
+{
+#ifdef Py_DEBUG
+    // assume there's a _PUSH_FRAME after every _CHECK_STACK_SPACE
+    bool expecting_push = false;
+    // if we ever can't get a framesize from _PUSH_FRAME, must exit soon
+    bool must_exit = false;
+#endif
+    int depth = 0;
+    uint32_t space_at_depth[MAX_ABSTRACT_FRAME_DEPTH];
+    uint32_t curr_space = 0;
+    uint32_t max_space = 0;
+    _PyUOpInstruction *first_valid_check_stack = NULL;
+    for (int pc = 0; pc < buffer_size; pc++) {
+        int opcode = buffer[pc].opcode;
+        switch (opcode) {
+            case _CHECK_STACK_SPACE: {
+#ifdef Py_DEBUG
+                assert(expecting_push == false);
+                assert(must_exit == false);
+                expecting_push = true;
+#endif
+                uint32_t framesize = 0;
+                if (pc + _CHECK_STACK_TO_PUSH_FRAME_OFFSET < buffer_size) {
+                    _PyUOpInstruction *uop = &buffer[
+                        pc + _CHECK_STACK_TO_PUSH_FRAME_OFFSET
+                    ];
+                    assert(uop->opcode == _PUSH_FRAME);
+                    PyCodeObject *code = get_co(uop);
+                    if (code != NULL) {
+                        framesize = code->co_framesize;
+                    }
+                }
+                if (framesize == 0) {
+#ifdef Py_DEBUG
+                    must_exit = true;
+#endif
+                    break;
+                }
+                if (first_valid_check_stack == NULL) {
+                    first_valid_check_stack = &buffer[pc];
+                }
+                else {
+                    // delete all but the first valid _CHECK_STACK_SPACE
+                    buffer[pc].opcode = _NOP;
+                }
+                space_at_depth[depth] = framesize;
+                curr_space += framesize;
+                max_space = curr_space > max_space ? curr_space : max_space;
+                depth++;
+                assert(depth < MAX_ABSTRACT_FRAME_DEPTH);
+                break;
+            }
+            case _POP_FRAME: {
+#ifdef Py_DEBUG
+                assert(expecting_push == false);
+                assert(must_exit == false);
+#endif
+                depth--;
+                assert(depth >= 0);
+                curr_space -= space_at_depth[depth];
+                break;
+            }
+#ifdef Py_DEBUG
+            case _PUSH_FRAME: {
+                assert(expecting_push == true);
+                expecting_push = false;
+                break;
+            }
+#endif
+            case _JUMP_TO_TOP:
+            case _EXIT_TRACE: {
+                if (first_valid_check_stack != NULL) {
+                    assert(first_valid_check_stack->opcode == _CHECK_STACK_SPACE);
+                    assert(max_space > 0);
+                    assert(max_space < INT_MAX);
```

Comment:
```suggestion
                    if (max_space < INT_MAX) {
```

Reply:
Maybe reduce the limit to `UNIT16_MAX`?

Reply:
Ah, setting a valid `_CHECK_STACK_SPACE_OPERAND` is mandatory if even a single `_CHECK_STACK_SPACE` is deleted during this pass, so I can't make this conditional on the value of `max_space`. I'm going to add some additional checks to ensure `max_space` is always an acceptable value and rework the exit logic